### ISSUE TITLE
Migrate away from monix to cats-effect

### ISSUE THREAD

**Issue by mcanlas** [scalacenter/bloop#2066]
Since `monix` is no longer maintained, we should upgrade the effect/cancelation framework to `cats-effect`

Hello from the Scala Spree at Scala Days 2023 Seattle!

**Comment by He-Pin** [comment]
Can it be done with zio or pekko?

**Comment by tgodzik** [comment]
zio probably also, but pekko would be an even major rewrite, so I would avoid that.

The biggest challenge is to probably use bsp4j first, later we can use any Framework we want underneath

**Comment by tgodzik** [comment]
bsp4s uses monix by default, so to change to cats we can either switch to bsp4j or rewrite bsp4s also to cats. I would prefer the first option since it's more framework agnostic and would be easier to change to something else if needed.

**Comment by KristianAN** [comment]
> bsp4s uses monix by default, so to change to cats we can either switch to bsp4j or rewrite bsp4s also to cats. I would prefer the first option since it's more framework agnostic and would be easier to change to something else if needed.

I have looked some at this and I'm not 100% sure if this is the best way to go.  Some points to consider

- There are some issues with using bsp4j. lsp4j does not have the best API, also it's based on CompletableFuture, which means we will have to convert between that and whatever effect system bloop uses/will use. 
- By migrating bsp4s/jsonrpc4s to fs2 (jsonrpc4s uses monix observables, of which I think fs2.Stream will be the most similar thing in the CE world) we can also support scala-native and scalaJS. 
- I don't think being framework agnostic is of importance here. Migrating from Monix to CE will be a huge undertaking. I doubt more migrations will happen (I really hope not!)
- By migrating jsonrpc4s/bsp4s to fs2 we can keep much of the API the same/similar. It should give an easier way of migration.

**Comment by tgodzik** [comment]
> There are some issues with using bsp4j. lsp4j does not have the best API, also it's based on CompletableFuture, which means we will have to convert between that and whatever effect system bloop uses/will use.

Yeah, but that should be pretty simple to do in the end, no?

But, migrating bsp4s is an additional option for sure. I haven't started with either now. I am thinking of doing it this year, maybe it would be simpler to start with bsp4s :thinking:

**Comment by KristianAN** [comment]
> Yeah, but that should be pretty simple to do in the end, no?
Yes, maybe? I don't know if cancellation semantics play a part here?
